### PR TITLE
Use GroupInfo instead of PublicGroupState. (2/2) 

### LINKED ISSUE #719: [MLS Spec change]  Refactor Group Joining and fix External Inits (#529)
**Description of the changes**
PGS and GroupInfo are joined into a single struct. Also fixes a big in External Commit that we had fixed locally before.

**Link to the exact changes**
https://github.com/mlswg/mls-protocol/pull/529

---

- [x] #1027

### LINKED ISSUE #1027: [MLS Spec change] Remove dangling references to PublicGroupState (#601)
**Link to the exact changes**
https://github.com/mlswg/mls-protocol/pull/601

**Description of the changes**
> PublicGroupState was removed as an independent struct in #529.  This PR fixes some dangling references to it.

---

Part of #719

### PR DESCRIPTION
**Pull request by duesee**
* Closes #1027
* Closes #719

This PR is very similar to the second half of #1106 but adds an additional `with_ratchet_tree` flag to `export_group_info`.

### CODE REVIEW

File: openmls/src/messages/mod.rs
Comment (franziskuskiefer) [inline openmls/src/messages/mod.rs:365]: Use `SIGNATURE_GROUP_INFO_LABEL` here as well.

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (franziskuskiefer) [inline openmls/src/group/core_group/new_from_external_init.rs:129]: I'm not sure this can ever happen.
Joining a group by external commit implies that there is a group already. The commit that is generated for this join moves the epoch forward. I therefore don't see how the epoch can be 0 here.
Is this actually used? Shouldn't we throw an error instead (if the above is correct)?

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (franziskuskiefer) [inline openmls/src/group/core_group/new_from_external_init.rs:140]: Throw a library error instead of unwrapping.

File: openmls/src/messages/tests/test_pgs.rs
Comment (franziskuskiefer) [inline openmls/src/messages/tests/test_pgs.rs:3]: This file should probably get a new name.

File: openmls/src/messages/tests/test_pgs.rs
Comment (franziskuskiefer) [inline openmls/src/messages/tests/test_pgs.rs:41]: We have the convention to use `expect` in tests for cases, at least where the checks happen.
This also serves as documentation for what the test is checking.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks. Generally looks good.
A few items and questions inlien.

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:287]: `into_verifiable_group_info` is used in tests, see [https://github.com/openmls/openmls/pull/1127/files#diff-d1d10ba424715e9e5529dc7069797c74a8d946448d844a9ab0bd994ab2e88352R113] or [https://github.com/openmls/openmls/pull/1127/files#diff-d1d10ba424715e9e5529dc7069797c74a8d946448d844a9ab0bd994ab2e88352R185].

What was done before in the `PublicGroupState` tests was to serialize and deserialize it. But I found this code to be distracting and (if it was for this reason)  (de)serialization should be tested more systematically elsewhere.

File: openmls/src/group/core_group/mod.rs
Comment (duesee) [reply openmls/src/group/core_group/mod.rs:558]: Yeah. That's generally something I would like to change over the time (but havn't done in this PR for now.) Whenever we need ownership of an object, we should let the caller decide how to obtain ownership and not take references and clone internally. https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control

Comment (duesee) [reply openmls/src/group/core_group/mod.rs:558]: The next PR in the pipeline will change that. https://github.com/openmls/openmls/pull/1116/files#diff-e7948aa030a226922756322ba4c0e6748caf5fcd80cf8dcd4d11cf92f3d73cb1R210 Do you still want to make the change now or can we wait for #1116 to land?

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:321]: I'm thinking if it would not be easier to open an issue for things like that and do a single PR that rectifies all of this close to some release. I will rectify it here, though. Just a thought.

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (duesee) [reply openmls/src/group/core_group/new_from_external_init.rs:129]: Without this condition, tests fail with a `InvalidCommit(ConfirmationTagMismatch)`. But, to be honest, I am not entirely sure the draft tells us what to do in this case.

In the failing tests, alice creates a group first (with no other members).

According to draft-16, the group should have ...

> Initialize a one-member group with the following initial values:
> [...]
> epoch: 0
> [...]
> Confirmed transcript hash: The zero-length octet string
> Interim transcript hash: The zero-length octet string

(In OpenMLS, the `interim_transcript_hash` is set to `b"\x00" * 32` and `confirmed_transcript_hash` is non-zero. I assume the `confirmed_transcript_hash` is correct.)

Then, alice exports a group state and bob joins by an external commit. Here, the epoch is 0 (and triggers the above condition).

Now, I am not sure, how the `interim_transcript_hash` should be calculated really. From my understanding of draft-16, I would expect the hash of a zero-length string (empty interim || empty confirmation). However, debugging the "old code" (with `PublicGroupState`) it seemed to me that `b""` was used instead. Given that all tests pass having this condition, I assumed it to be fine.

Could you shed some light on this`

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:287]: Sorry, I still don't understand. From my understanding this is a method available throughout the crate during testing only. And it should be only available during tests, because, in real life, clients will deserialize a `VerifiableGroupInfo` and not create it from a `GroupInfo`.

> Is there an integration test covering this? Not sure if I've seen one.

Sorry, again. I don't understand what "this" refers to.

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (duesee) [reply openmls/src/group/core_group/new_from_external_init.rs:129]: I've created https://github.com/openmls/openmls/issues/1129.

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:287]: As discussed OOB, I've added an integration test to see how a member can join by using `export_group_info`, and `join_by_external_commit`.

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (franziskuskiefer) [reply openmls/src/group/core_group/new_from_external_init.rs:129]: Can you link the issue here in the code as well?

File: openmls/src/messages/mod.rs
Comment (franziskuskiefer) [inline openmls/src/messages/mod.rs:344]: Accessing values in the verifable object is always a little tricky because the idea is of course that you shouldn't get access until it's verified.
But in many cases, like this one, you need to get to some values in order to actually verify.
Can you add the reason for each of these getters why they are needed so no one has to guess in the future?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks for the test!
Generally looks good to me. Just a few nits, then this is good to go.

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:344]: Very good point. Maybe we can also go a step further and make it even more explicit by using a `_unverified` suffix for these?

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (duesee) [reply openmls/src/group/core_group/new_from_external_init.rs:74]: Type inference :P

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:344]: Ping @franziskuskiefer. (Will re-request a review later.)

File: openmls/src/group/core_group/new_from_external_init.rs
Comment (franziskuskiefer) [reply openmls/src/group/core_group/new_from_external_init.rs:74]: `verifiable_group_info.verify` works just fine 🙂

File: openmls/src/messages/mod.rs
Comment (duesee) [reply openmls/src/messages/mod.rs:344]: Alright :+1:

**Review by franziskuskiefer — APPROVED** [review]
Thanks, lgtm.

> Rebase?

I'd just squash and merge.

**Comment by duesee** [comment]
Sorry, had to reformat the code which dismissed the review.

**Review by franziskuskiefer — APPROVED** [review]

**Review by franziskuskiefer — APPROVED** [review]

**Pull request merged**
